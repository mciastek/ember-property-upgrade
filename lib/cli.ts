#!/usr/bin/env node

import fs from 'fs';
import path from 'path';
import readline from 'readline';
import util from 'util';

import chalk from 'chalk';
import program from 'commander';
import glob from 'glob';

import packageInfo from '../package.json';
import { Options, transform } from './index';

// tslint:disable-next-line no-empty
const noop = () => {};

program
  .version(packageInfo.version, '-v, --version')
  .description(`
    CLI for migration to new computed properties in Ember
    Usage: ember-property-upgrade [filesGlob] [options]
  `)
  .option('--no-format', 'Disable auto formatting after code parsing')
  .option('--prettier-config-file [value]', 'Path to Prettier config file')
  .option('--prettier-config [value]', 'Prettier config as JSON string')
  .option(
    '--framework-pkg [value]',
    'Name of Ember\'s import alias',
    noop,
    'Ember',
  )
  .option(
    '--computed-fn-name [value]',
    'Name of computed property function name',
    noop,
    'computed',
  )
  .parse(process.argv);

if (!process.argv.slice(2).length) {
  program.outputHelp();
}

const readFile = util.promisify(fs.readFile);
const writeFile = util.promisify(fs.writeFile);
const [filesGlob] = program.args;

const getPrettierConfigFromFile = (filePath: string) => {
  const resolvedPath = path.resolve(__dirname, filePath);

  if (fs.existsSync(resolvedPath)) {
    const result = require(resolvedPath);

    return result;
  }

  console.warn(chalk.yellow(`Config file "${filePath}" doesn't exist!`));

  return {};
};

const getPrettierConfig = (optionsString?: string, filePath?: string) => {
  if (optionsString) {
    try {
      return JSON.stringify(optionsString);
    } catch (error) {
      console.warn(chalk.yellow(`Parsing options string failed`));
      console.error(error);
    }
  }

  if (filePath) {
    return getPrettierConfigFromFile(filePath);
  }

  return {};
};

const transformOptions: Options = {
  packageName: program.frameworkPkg,
  computedFunName: program.computedFnName,
  autoFormat: !program.noFormat,
  autoFormatOptions: getPrettierConfig(
    program.prettierConfig,
    program.prettierConfigFile,
  ),
};

const drawProgress = (current: number, total: number) => {
  readline.clearLine(process.stdout, 0);
  readline.cursorTo(process.stdout, 0);
  process.stdout.write(`Processing... ${current}/${total}`);
};

const getParsedFile = async (filePath: string) => {
  try {
    const code = await readFile(filePath, { encoding: 'utf8' });
    const result = transform(code, transformOptions);

    await writeFile(filePath, result, { encoding: 'utf8' });
  } catch (error) {
    throw error;
  }
};

const parseFiles = (matches: string[]) => {
  let current = 1;

  return matches.map((filePath) => {
    getParsedFile(filePath)
      .then(() => {
        drawProgress(current, matches.length);
        current += 1;

        return filePath;
      });
  });
};

glob(filesGlob, (err, matches) => {
  if (err) {
    // tslint:disable-next-line no-console
    console.log(chalk.red('Something went wrong!'));

    return console.error(err);
  }

  Promise.all(parseFiles(matches))
    .then(() => {
      // tslint:disable-next-line no-console
      console.log(chalk.green('All done!'));
      process.exit(0);
    })
    .catch((error: Error) => {
      // tslint:disable-next-line no-console
      console.log(chalk.red('Something went wrong!'));

      return console.error(error);
    });
});
