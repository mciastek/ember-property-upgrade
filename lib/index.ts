import generate from '@babel/generator';

import {
  parse,
  ParserOptions,
} from '@babel/parser';

import traverse, { NodePath } from '@babel/traverse';
import * as t from '@babel/types';

const DEFAULT_OPTIONS: ParserOptions = {
  sourceType: 'module',
};

export const transform = (input: string) => {
  const ast = parse(input, DEFAULT_OPTIONS);

  let callExpression: t.CallExpression;
  let propertyArgs;

  traverse(ast, {
    Identifier(path: NodePath<t.Identifier>) {
      if (path.node.name === 'property') {
        callExpression = path.parentPath.parent as t.CallExpression;
        propertyArgs = callExpression.arguments;

        path.parentPath.parentPath.replaceWith(t.callExpression(
          t.memberExpression(t.identifier('Ember'), t.identifier('computed')),
          [
            ...propertyArgs,
            callExpression.callee.object,
          ],
        ));
      }
    },
  });

  const { code } = generate(ast);

  return code;
};
