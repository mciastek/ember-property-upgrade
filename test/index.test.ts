import { getFixtureFile } from './helpers';

import { transform } from 'lib/index';

describe('ember-property-upgrade', () => {
  let input: string;

  describe('for basic examples', () => {
    beforeAll(() => {
      input = getFixtureFile('basic-example.js');
    });

    it('transforms input', () => {
      const result = transform(input);

      expect(result).toBeDefined();
      expect(result).toMatchSnapshot();
    });
  });

  describe('for complex examples', () => {
    beforeAll(() => {
      input = getFixtureFile('complex-example.js');
    });

    it('transforms input', () => {
      const result = transform(input);

      expect(result).toBeDefined();
      expect(result).toMatchSnapshot();
    });
  });

  describe('for array examples', () => {
    beforeAll(() => {
      input = getFixtureFile('array-example.js');
    });

    it('transforms input', () => {
      const result = transform(input);

      expect(result).toBeDefined();
      expect(result).toMatchSnapshot();
    });
  });

  describe('for invalid examples', () => {
    beforeAll(() => {
      input = getFixtureFile('invalid-example.js');
    });

    it('transforms input', () => {
      const result = transform(input);

      expect(result).toBeDefined();
      expect(result).toMatchSnapshot();
    });
  });

  describe('for no transform examples', () => {
    beforeAll(() => {
      input = getFixtureFile('no-transform-example.js');
    });

    it('return empty output', () => {
      const result = transform(input);

      expect(result).toBeDefined();
      expect(result).toMatchSnapshot();
    });
  });
});
