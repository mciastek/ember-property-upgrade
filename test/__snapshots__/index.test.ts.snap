// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ember-property-upgrade for basic examples transforms input 1`] = `
"const Person = EmberObject.extend({
  fullName: Ember.computed('firstName', 'lastName', function () {
    return \`\${this.firstName} \${this.lastName}\`;
  })
});"
`;

exports[`ember-property-upgrade for complex examples transforms input 1`] = `
"import EmberObject, { computed } from '@ember/object';
Person = EmberObject.extend({
  // these will be supplied by \`create\`
  firstName: null,
  lastName: null,
  powers: [],
  fullName: Ember.computed('firstName', 'lastName', function () {
    return \`\${this.firstName} \${this.lastName}\`;
  }),
  mainPower: Ember.computed('powers', function () {
    return this.powers[0];
  })
});
let ironMan = Person.create({
  firstName: 'Tony',
  lastName: 'Stark',
  powers: ['super-smart']
});
ironMan.fullName; // \\"Tony Stark\\""
`;
